### PR TITLE
Timer units

### LINKED ISSUE #10: Exporting a timer with seconds isn't collecting in the right units
Ran into an interesting issue when attempting to track request duration via:

```
let start = Date()
let oneSecond = -start.timeIntervalSince(start.addingTimeInterval(1.0))
```

It resulted in much larger values than I expected, to show some sample test output comparing the unexpected values:

```
error: -[SwiftPrometheusTests.PrometheusMetricsTests testSummaryDouble] : XCTAssertEqual failed: ("# TYPE summary_seconds summary
summary_seconds{quantile="0.01"} 1000000000.0
summary_seconds{quantile="0.05"} 1000000000.0
summary_seconds{quantile="0.5"} 4000000000.0
summary_seconds{quantile="0.9"} 10000000000000.0
summary_seconds{quantile="0.95"} 10000000000000.0
summary_seconds{quantile="0.99"} 10000000000000.0
summary_seconds{quantile="0.999"} 10000000000000.0
summary_seconds_count 5
summary_seconds_sum 10130000000000
summary_seconds{quantile="0.01", myValue="labels"} 123000000000.0
summary_seconds{quantile="0.05", myValue="labels"} 123000000000.0
summary_seconds{quantile="0.5", myValue="labels"} 123000000000.0
summary_seconds{quantile="0.9", myValue="labels"} 123000000000.0
summary_seconds{quantile="0.95", myValue="labels"} 123000000000.0
summary_seconds{quantile="0.99", myValue="labels"} 123000000000.0
summary_seconds{quantile="0.999", myValue="labels"} 123000000000.0
summary_seconds_count{myValue="labels"} 1
summary_seconds_sum{myValue="labels"} 123000000000") 

is not equal to

("# TYPE summary_seconds summary
summary_seconds{quantile="0.01"} 1.0
summary_seconds{quantile="0.05"} 1.0
summary_seconds{quantile="0.5"} 4.0
summary_seconds{quantile="0.9"} 10000.0
summary_seconds{quantile="0.95"} 10000.0
summary_seconds{quantile="0.99"} 10000.0
summary_seconds{quantile="0.999"} 10000.0
summary_seconds_count 5
summary_seconds_sum 10130
summary_seconds{quantile="0.01", myValue="labels"} 123.0
summary_seconds{quantile="0.05", myValue="labels"} 123.0
summary_seconds{quantile="0.5", myValue="labels"} 123.0
summary_seconds{quantile="0.9", myValue="labels"} 123.0
summary_seconds{quantile="0.95", myValue="labels"} 123.0
summary_seconds{quantile="0.99", myValue="labels"} 123.0
summary_seconds{quantile="0.999", myValue="labels"} 123.0
summary_seconds_count{myValue="labels"} 1
summary_seconds_sum{myValue="labels"} 123")
```

The underlying `swift-metrics` saves everything as `Int64` and [assumes nanoseconds](https://github.com/apple/swift-metrics/blob/master/Sources/CoreMetrics/Metrics.swift#L220), as all other methods convert then call to that method. From the [Prometheus docs on naming](https://prometheus.io/docs/instrumenting/writing_exporters/#naming):

> Metrics must use base units (e.g. seconds, bytes) and leave converting them to something more readable to graphing tools. No matter what units you end up using, the units in the metric name must match the units in use.

This follows [an assumption that Prometheus client libraries use seconds](https://www.robustperception.io/who-wants-seconds) as well.

I believe we should strive to create an idiomatic client, which I think means exporting the metrics in the desired unit. Given the assumption that units are encoded in the metric name, I think that's potentially a good path to investigate further.

### PR DESCRIPTION
**Pull request by MrLotU**
Fixes #10 

### Checklist
- [x] The provided tests still run.
- [x] I've created new tests where needed.
- [x] I've updated the documentation if necessary.

This behavior is only available (and required) for Summaries created using `swift-metrics`.

### CODE REVIEW

File: Tests/SwiftPrometheusTests/PrometheusMetricsTests.swift [hunk 1]
```diff
@@ -146,6 +146,31 @@ final class PrometheusMetricsTests: XCTestCase {
         """)
     }
     
+    func testSummaryWithDisplayUnit() {
```
Comment (ktoso) [inline Tests/SwiftPrometheusTests/PrometheusMetricsTests.swift]: `...WithPreferredDisplayUnit`?

File: Tests/SwiftPrometheusTests/PrometheusMetricsTests.swift [hunk 2]
```diff
@@ -117,8 +117,8 @@ final class PrometheusMetricsTests: XCTestCase {
         summary.recordNanoseconds(4)
         summary.recordNanoseconds(10000)
         
-        let summaryTwo = Timer(label: "my_summary", dimensions: [("myValue", "labels")])
-        summaryTwo.recordNanoseconds(123)
+//        let summaryTwo = Timer(label: "my_summary", dimensions: [("myValue", "labels")])
+//        summaryTwo.recordNanoseconds(123)
```
[the remark this answers is not in the document]
Comment (MrLotU) [reply Tests/SwiftPrometheusTests/PrometheusMetricsTests.swift]: Yeah, I commented that temporarily. There seems to be a bug with my summaries / histograms that I was checking out. I have it un commented on my local machine, just not pushed :)

File: Tests/SwiftPrometheusTests/PrometheusMetricsTests.swift (hunk 1 above)
Comment (MrLotU) [reply Tests/SwiftPrometheusTests/PrometheusMetricsTests.swift]: 👍

File: Package.resolved [hunk 3]
```diff
@@ -15,8 +15,8 @@
         "repositoryURL": "https://github.com/apple/swift-nio.git",
         "state": {
           "branch": null,
-          "revision": "9201908b54578aa33f1d1826a5a680aca8991843",
-          "version": "2.8.0"
+          "revision": "6a79d4687ede7c088fe3e31d9b3452d6e6cd91eb",
+          "version": "2.5.1"
```
[the remark this answers is not in the document]
Comment (MrLotU) [reply Package.resolved]: Nope, local Xcode cache doing its thing. I'll fix this :)

**Comment by MrLotU** [comment]
@Yasumoto Since you opened #10, would love to have your feedback on this one.

File: Tests/SwiftPrometheusTests/PrometheusMetricsTests.swift [hunk 4]
```diff
@@ -142,7 +142,32 @@ final class PrometheusMetricsTests: XCTestCase {
...
-        my_summary_sum{myValue="labels"} 123
+        my_summary_sum{myValue="labels"} 123.0
+        """)
+    }
+    
+    func testSummaryWithPreferredDisplayUnit() {
+        let summary = Timer(label: "my_summary", preferredDisplayUnit: .seconds)
+        
+        summary.recordSeconds(1)
+        summary.recordSeconds(2)
```
Comment (Yasumoto) [inline Tests/SwiftPrometheusTests/PrometheusMetricsTests.swift]:
```suggestion
        summary.recordMilliseconds(2 * 1_000)
```

File: Tests/SwiftPrometheusTests/PrometheusMetricsTests.swift [hunk 5]
```diff
@@ -142,7 +142,32 @@ final class PrometheusMetricsTests: XCTestCase {
...
+        summary.recordSeconds(4)
```
Comment (Yasumoto) [inline Tests/SwiftPrometheusTests/PrometheusMetricsTests.swift]:
```suggestion
        summary.recordNanoseconds(4 * 1_000_000_000)
```

I think it'd be nice to also exercise other units to make sure we don't regress in the future.

**Comment by Yasumoto** [comment]
Ah, one last note, any chance you could run `swift test --generate-linuxmain` ? That might be nice to tie into CI so we don't forget it too.

```
jmsmith@SFO-M-JMSMITH02 ~/w/g/M/SwiftPrometheus (TimerUnits)> git diff                 19:48:58
diff --git a/Tests/SwiftPrometheusTests/XCTestManifests.swift b/Tests/SwiftPrometheusTests/XCTestManifests.swift
index 7862bde..600f402 100644
--- a/Tests/SwiftPrometheusTests/XCTestManifests.swift
+++ b/Tests/SwiftPrometheusTests/XCTestManifests.swift
@@ -14,6 +14,7 @@ extension PrometheusMetricsTests {
         ("testHistogram", testHistogram),
         ("testMetricDestroying", testMetricDestroying),
         ("testSummary", testSummary),
+        ("testSummaryWithPreferredDisplayUnit", testSummaryWithPreferredDisplayUnit),
     ]
 }

@@ -25,6 +26,7 @@ extension SwiftPrometheusTests {
         ("testCounter", testCounter),
         ("testGauge", testGauge),
         ("testHistogram", testHistogram),
+        ("testMultipleCounter", testMultipleCounter),
         ("testSummary", testSummary),
     ]
 }
```

**Comment by Yasumoto** [comment]
So pumped this is out there!! Thanks for all the help and hard work, @MrLotU! 🎉

**Pull request merged**